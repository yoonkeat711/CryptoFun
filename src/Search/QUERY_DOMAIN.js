const QUERY_DOMAIN = 'https://api.coingecko.com/api/v3';

export default QUERY_DOMAIN;
