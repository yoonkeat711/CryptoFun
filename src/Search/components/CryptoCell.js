import React  from 'react';
import {View, Text, StyleSheet} from 'react-native';

const CryptoCell = ({symbol, name, index}) => {
  return (
    <View
      style={{
        flex: 1,
        flexDirection: 'row',
        backgroundColor: index % 2 === 1 ? 'white' : '#F0F8FF',
        width: '100%',
        justifyContent: 'space-between',
        paddingHorizontal: 16,
        paddingVertical: 10,
      }}>
      <Text style={{width: 50}}>{index}</Text>
      <Text style={{flex: 1, textAlign: 'left'}} numberOfLines={1}>
        {symbol.toUpperCase()}
      </Text>
      <Text style={{flex: 3, textAlign: 'left'}} numberOfLines={1}>
        {name}
      </Text>
    </View>
  );
};

export default CryptoCell;
