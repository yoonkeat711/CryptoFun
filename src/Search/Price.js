import React, {useState, useEffect, Component, PureComponent} from 'react';
import {
  View,
  Text,
  TextInput,
  ActivityIndicator,
  TouchableOpacity,
  AsyncStorage,
} from 'react-native';
import QUERY_DOMAIN from './QUERY_DOMAIN';

// const Price = ({route, navigation}) => {
//
//   const [price, setPrice] = useState(0);
//
//   React.useEffect = (() => {
//     const PRICE_API =
//       `${QUERY_DOMAIN}` +
//       '/simple/price?ids=' +
//       `${route.params?.item?.id}` +
//       '&vs_currencies=myr';
//   //     // fetch(PRICE_API, {
//   //     //   method: 'GET',
//   //     //   headers: {
//   //     //     Accept: 'application/json',
//   //     //     'Content-Type': 'application/json',
//   //     //   },
//   //     // })
//   //     //   .then((response) => response.json())
//   //     //   .then((response) => {
//   //     //     console.warn(response);
//   //     //     setPrice(response);
//   //     //   })
//   //     //   .catch((err) => {
//   //     //     console.warn(err);
//   //     //   });
//     }, []);
//
//   const fetchPrice = React.useCallback(async() => {
//     const response = await fetch(PRICE_API);
//
//   },[]);
//
//   return (
//     <View style={{flex: 1}}>{console.warn(price)}</View>
//   );
// };

class LabelValue extends PureComponent {
  constructor(props) {
    super(props);
  }

  render() {
    const {label, value} = this.props;

    return (
      <View
        style={{
          flexDirection: 'row',
          justifyContent: 'space-between',
          paddingHorizontal: 16,
          paddingVertical: 8,
        }}>
        <Text style={{fontWeight: '700'}}>{label}</Text>
        <Text>{value}</Text>
      </View>
    );
  }
}

const cartStorageKey = '@storage:cart';
const historyStorageKey = '@storage:history';

class Price extends Component {
  constructor(props) {
    super(props);

    this.state = {
      price: 0,
      totalPrice: 0,
      unit: 1,
      isLoading: true,
    };
  }

  async componentDidMount(): void {
    const PRICE_API =
      `${QUERY_DOMAIN}` +
      '/simple/price?ids=' +
      `${this.props.route.params?.item?.id}` +
      '&vs_currencies=myr';

    fetch(PRICE_API)
      .then((response) => response.json())
      .then((response) => {
        const price = response[this.props.route.params?.item?.id];
        this.setState({
          price: price['myr'],
          isLoading: false,
        });
      });
  }

  onAmountChange = (value) => {
    this.setState({
      unit: value,
      totalPrice: value * this.state.price,
    });
  };

  onPressCheckout = async (value) => {
    // await AsyncStorage.setItem(historyStorageKey, value);
  };

  render() {
    const {id, name, symbol} = this.props.route.params?.item;

    if (this.state.isLoading) {
      return <ActivityIndicator style={{flex: 1, alignSelf: 'center'}} />;
    } else
      return (
        <View style={{backgroundColor: 'white', flex: 1}}>
          <LabelValue label={'NAME'} value={name} />
          <LabelValue label={'SYMBOL'} value={symbol.toUpperCase()} />
          <LabelValue label={'CURRENCY'} value={'MYR'} />
          <LabelValue label={'VALUE'} value={this.state.price} />
          <View
            style={{
              flexDirection: 'row',
              justifyContent: 'space-between',
              paddingHorizontal: 16,
              paddingVertical: 8,
            }}>
            <Text style={{fontWeight: '700'}}>{'UNIT'}</Text>
            <TextInput
              style={{
                borderWidth: 1,
                width: 100,
                height: 35,
                textAlign: 'right',
              }}
              value={this.state.unit}
              onChangeText={(value) => this.onAmountChange(value)}
              keyboardType={'numeric'}
            />
          </View>
          <View
            style={{height: 1, backgroundColor: 'grey', marginHorizontal: 16}}
          />
          <View
            style={{
              flexDirection: 'row',
              justifyContent: 'space-between',
              paddingTop: 8,
              paddingHorizontal: 16,
            }}>
            <Text style={{fontWeight: '700'}}>{'TOTAL'}</Text>
            <Text>{this.state.totalPrice.toFixed(2)}</Text>
          </View>
          <View
            style={{
              width: '100%',
              justifyContent: 'space-between',
              flexDirection: 'row',
              paddingHorizontal: 36,
              paddingVertical: 16,
              position: 'absolute',
              bottom: 0,
              flex: 1,
              borderTopWidth: 0.5,
            }}>
            <TouchableOpacity style={{}} onPress={() => this.onPressCheckout()}>
              <Text>{'Checkout'}</Text>
            </TouchableOpacity>
            <TouchableOpacity>
              <Text>{'Add to Cart'}</Text>
            </TouchableOpacity>
          </View>
        </View>
      );
  }
}

export default Price;
