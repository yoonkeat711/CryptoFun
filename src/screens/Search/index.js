import React, {Component, useEffect, useState, useCallback} from 'react';
import {
  View,
  Text,
  SafeAreaView,
  FlatList,
  TextInput,
  TouchableOpacity,
} from 'react-native';
import CryptoCell from './components/CryptoCell';
import MockCoinList from './coinlist.json';
import QUERY_DOMAIN from './QUERY_DOMAIN';
import store from '../../store';
import {displayCart} from '../Cart/actions';
import AsyncStorage from '@react-native-community/async-storage';
import storageKey from '../../constants/storageKey';
import {displayHistory} from '../History/actions';

const COIN_LIST_API = `${QUERY_DOMAIN}` + '/coins/list';

const Search = ({navigation}) => {
  const [data, setData] = useState({});
  const [loading, setLoading] = useState(false);
  const [text, setText] = useState(null);

  const fetchData = useCallback(() => {
    // fetch(COIN_LIST_API, {
    //   method: 'GET',
    //   headers: {
    //     Accept: 'application/json',
    //     'Content-Type': 'application/json',
    //   },
    // })
    //   .then((response) => response.json())
    //   .then((response) => {
    //     setData(response);
    //   })
    //   .catch((err) => {
    //     console.warn(err);
    //   });
    setData(MockCoinList);
  }, []);

  useEffect(async() => {
    fetchData();
    const cartItemStorage = JSON.parse(await AsyncStorage.getItem(storageKey.cart));
    const historyItemStorage = JSON.parse(await AsyncStorage.getItem(storageKey.history));

    store.dispatch(displayCart(cartItemStorage));
    store.dispatch(displayHistory(historyItemStorage));
  }, [fetchData]);

  const filteredData = (() => {
    if (!text?.trim().length) {
      return data;
    }

    return data.filter((item) => {
      return (
        item.name.toLowerCase().includes(text.toLowerCase()) ||
        item.symbol.toLowerCase().includes(text.toLowerCase())
      );
    });
  })();

  const navigateToPrice = (item) => {
    navigation.navigate('Price', {
      item: item,
    });
  }

  return (
    <View>
      <SafeAreaView>
        <TextInput
          placeholder={'Search currency here or currency symbol here'}
          style={{marginHorizontal: 16, paddingVertical: 8}}
          onChangeText={(value) => setText(value)}
        />
        <FlatList
          data={filteredData}
          initialNumToRender={5}
          renderItem={({item, index}) => (
            <TouchableOpacity onPress={() => navigateToPrice(item)}>
            <CryptoCell
              symbol={item.symbol}
              name={item.name}
              index={index + 1}
            />
            </TouchableOpacity>
          )}
          ListHeaderComponent={() => (
            <CryptoCell symbol={'SYMBOL'} name={'NAME'} index={'#'} />
          )}
          ListHeaderComponentStyle={{
            borderBottomWidth: 0.5,
            borderBottomColor: 'grey',
          }}
          ItemSeparatorComponent={() => {
            return <View style={{height: 0.5, backgroundColor: 'grey'}} />;
          }}
        />
      </SafeAreaView>
    </View>
  );
};


export default Search;
