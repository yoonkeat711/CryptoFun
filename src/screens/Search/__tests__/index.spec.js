import Search from '../index';
import renderer from 'react-test-renderer';
import React from 'react';
import CryptoCell from '../components/CryptoCell';

test('renders correctly', () => {
  const tree = renderer.create(<Search />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('renders CryptoCell correctly', () => {
  const renderTree = renderer.create(<CryptoCell symbol={"BTC"} name={"Bitcoin"} index={1} />);
  const renderInstance = renderTree.root;
  console.log(renderInstance);
});
