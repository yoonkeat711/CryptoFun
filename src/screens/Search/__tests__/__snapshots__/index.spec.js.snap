// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Search currency here or currency symbol here"
      rejectResponderTermination={true}
      style={
        Object {
          "marginHorizontal": 16,
          "paddingVertical": 8,
        }
      }
      underlineColorAndroid="transparent"
    />
    <RCTScrollView
      ItemSeparatorComponent={[Function]}
      ListHeaderComponent={[Function]}
      ListHeaderComponentStyle={
        Object {
          "borderBottomColor": "grey",
          "borderBottomWidth": 0.5,
        }
      }
      data={null}
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={5}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View>
        <View
          onLayout={[Function]}
          style={
            Object {
              "borderBottomColor": "grey",
              "borderBottomWidth": 0.5,
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "#F0F8FF",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "space-between",
                "paddingHorizontal": 16,
                "paddingVertical": 10,
                "width": "100%",
              }
            }
          >
            <Text
              style={
                Object {
                  "width": 50,
                }
              }
            >
              #
            </Text>
            <Text
              numberOfLines={1}
              style={
                Object {
                  "flex": 1,
                  "textAlign": "left",
                }
              }
            >
              SYMBOL
            </Text>
            <Text
              numberOfLines={1}
              style={
                Object {
                  "flex": 3,
                  "textAlign": "left",
                }
              }
            >
              NAME
            </Text>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </RCTSafeAreaView>
</View>
`;
